website/project/model.py

```diff
@@ -2825,149 +2809,117 @@ def to_json(self):
-    #     'disapproval_token': 'UotzClTFOic2PYxHDStby94bCQMwJy'}
+    #     'approval_token': 'Pew7wj1Puf7DENUPFPnXSwa1rf3xPN',
+    #     'disapproval_token': 'TwozClTFOic2PYxHDStby94bCQMwJy'}
     # }
     approval_state = fields.DictionaryField()
-    # One of 'pending', 'retracted', or 'cancelled'
-    state = fields.StringField(default='pending', validate=validate_retraction_state)
+    # One of 'unapproved', 'active', 'cancelled', or 'completed
+    state = fields.StringField(default='unapproved', validate=validate_sanction_state)
 
     def __repr__(self):
-        parent_registration = Node.find_one(Q('retraction', 'eq', self))
-        return ('<Retraction(parent_registration={0}, initiated_by={1}) '
-                'with _id {2}>').format(
-            parent_registration,
+        return '<Sanction(initiated_by={0}, end_date={1}) with _id {2}>'.format(
             self.initiated_by,
+            self.end_date,
             self._id
         )
 
     @property
-    def is_retracted(self):
-        return self.state == self.RETRACTED
+    def pending_approval(self):
+        return self.state == Sanction.UNAPPROVED
 
-    @property
-    def pending_retraction(self):
-        return self.state == self.PENDING
+    def _validate_authorizer(self, user):
+        return True
 
-    def disapprove_retraction(self, user, token):
-        """Cancels retraction if user is admin and token verifies."""
-        try:
-            if self.approval_state[user._id]['disapproval_token'] != token:
-                raise InvalidRetractionDisapprovalToken('Invalid retraction disapproval token provided.')
-        except KeyError:
-            raise PermissionsError('User must be an admin to disapprove retraction of a registration.')
+    def add_authorizer(self, user, approved=False, save=False):
+        self._validate_authorizer(user)
+        if user._id not in self.approval_state:
+            self.approval_state[user._id] = {
+                'has_approved': approved,
+                'approval_token': security.random_string(30),
+                'disapproval_token': security.random_string(30),
+            }
+            if save:
+                self.save()
+            return True
+        return False
 
-        self.state = self.CANCELLED
-        parent_registration = Node.find_one(Q('retraction', 'eq', self))
-        parent_registration.registered_from.add_log(
-            action=NodeLog.RETRACTION_CANCELLED,
-            params={
-                'node': parent_registration._id,
-                'retraction_id': self._id,
-            },
-            auth=Auth(user),
-            save=True,
-        )
+    def remove_authorizer(self, user):
+        if user._id in self.approval_state:
+            del self.approval_state[user._id]
+            self.save()
+            return True
+        return False
 
-    def approve_retraction(self, user, token):
+    def _on_complete(self, user):
+        pass
```

Comment:
There are 3 events we want to capture:
- rejection (any rejection)
- approval (any approval)
- completion (unanimous approval) 
And perhaps the 4th event: failure

Imagine the use case where unanimous rejection is necessary to reject a Sanction?

So maybe 5 callbacks are needed:
- #_on_approve (one user approves)
- #_on_approval (unanimous approval)
- #_on_reject (one user rejects)
- #_on_rejection (unanimous rejection)
- #_on_complete (everyone has responded)